lang/cs/Org.Apache.REEF.Examples.PeerToPeer/Communication/MessageStreamingCodec.cs

```diff
@@ -0,0 +1,172 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Wake.Remote;
+using Org.Apache.REEF.Wake.StreamingCodec;
+
+namespace Org.Apache.REEF.Examples.PeerToPeer.Communication
+{
+    /// <summary>
+    /// Streaming Codec for the Group Communication Message
+    /// </summary>
+    internal sealed class MessageStreamingCodec<T> : IStreamingCodec<Message<T>>
+    {
+        private readonly IStreamingCodec<T> _codec;
+
+        /// <summary>
+        /// Empty constructor to allow instantiation by reflection
+        /// </summary>
+        [Inject]
+        private MessageStreamingCodec(IStreamingCodec<T> codec)
+        {
+            _codec = codec;
+        }
+
+        /// <summary>
+        /// Read the class fields.
+        /// </summary>
+        /// <param name="reader">The reader from which to read </param>
+        /// <returns>The Group Communication Message</returns>
+        public Message<T> Read(IDataReader reader)
+        {
+            int metadataSize = reader.ReadInt32();
+            byte[] metadata = new byte[metadataSize];
+            reader.Read(ref metadata, 0, metadataSize);
+            var res = GenerateMetaDataDecoding(metadata);
+
+            string source = res.Item1;
+            string destination = res.Item2;
+
+            T data = _codec.Read(reader);
+            if (data == null)
+            {
+                throw new Exception(
+                    "message instance cannot be created from the IDataReader in Group Communication Message");
+            }
+
+            return new Message<T>(source, destination, data);
+        }
+
+        /// <summary>
+        /// Writes the class fields.
+        /// </summary>
+        /// <param name="obj">The message to write</param>
+        /// <param name="writer">The writer to which to write</param>
+        public void Write(Message<T> obj, IDataWriter writer)
+        {
+            byte[] encodedMetadata = GenerateMetaDataEncoding(obj);
+            byte[] encodedInt = BitConverter.GetBytes(encodedMetadata.Length);
+            byte[] totalEncoding = encodedInt.Concat(encodedMetadata).ToArray();
+            writer.Write(totalEncoding, 0, totalEncoding.Length);
+
+            _codec.Write(obj.Data, writer);
```

Comment:
Can't we just use the codec for conversion of objects into bytes? e.g. simply write `_codec.Write(obj, writer)`?